Remove redundant eunit tests

raghavkarol opened the pull request:
These were the original tests writen by lixen and all of these cases
are covered in the ../test/riak_kv_sweeper_SUITE.erl These tests, do
not run stable on `buildbot`; at least two separate and obscure
failures have been observed, the meck process dies or eunit times out
trying to run the tests.

Commit:
Remove redundant eunit tests

These were the original tests writen by lixen and all of these cases
are covered in the ../test/riak_kv_sweeper_SUITE.erl These tests, do
not run stable on `buildbot`; at least two separate and obscure
failures have been observed, the meck process dies or eunit times out
trying to run the tests.

JeetKunDoug commented:
+1 - I'd like to see some additional comments in the replacement SUITE as to what the tests are testing, and how, exactly, they prove what they test though.

Commit:
Adding comments explaining tests cases and what they assert

Documenting the test cases was a good sanity check of their
implementation. The following test cases were improved as a
result.

- test_sweep_request_non_existing_index :: added match on return value

- test_scheduler_run_interval :: refactored

- test_scheduler_async_backend :: new explicity test `async` backend
  behaviour, see test_scheduler_sync_backend

- test_scheduler_run_interval_as_fun :: new test

- test_scheduler_queue :: added match to assert that sweep_queue_time
changes

- test_scheduler_sweep_window_never :: using `sync` backend to avoid
race condition

- test_scheduler_now_outside_sleep_window :: using `sync` backedn to
avoid race condtions

- test_stop_all_scheduled_sweeps :: add match to ensure we have a
running sweep

- test_scheduler_estimated_keys_lock_ok :: add match to check the
correct value of EstimatedKeys is passed.

- test_scheduler_estimated_keys_lock_fail :: add match to check the
correct value of EstimatedKeys is passed.

raghavkarol commented:
@JeetKunDoug updated the replacement suite as requested.

nickelization commented on test/riak_kv_sweeper_SUITE.erl:
```diff
@@ -234,10 +382,21 @@ test_scheduler_remove_participant(Config) ->
     Indices = ?config(vnode_indices, Config),
     new_sweep_participant(sweeper_callback_wait),
-    sweeper_meck_helper:backend_modules(async, _NumKeys=5000, _ObjecSizeBytes=0),
+    sweeper_meck_helper:backend_modules(async, _NumKeys=1000, _ObjecSizeBytes=0),
```
I just noticed we misspell "Object" as "Objec" in a bunch of places all over this test suite.

nickelization commented on test/riak_kv_sweeper_SUITE.erl:
```diff
@@ -207,16 +309,62 @@ test_scheduler_worker_process_crashed(Config) ->
+    sweeper_meck_helper:backend_modules(async, _NumKeys=1000, _ObjecSizeBytes=0),
+
+    sweep_all_indices(Indices),
+    timer:sleep(timer:seconds(RunInterval + 1)),
```
I know it's a short sleep, but since we're just adding this as a new test, is there a way we could rework the test to eliminate it? Partly I'd like the test to run as fast as possible, but I'm also slightly concerned as to whether this will pass reliably - it seems like it should, but we've seen a lot of wonky timing-related failures on the builders lately...

raghavkarol replied:
This is not a new test, just looks like that in the diff. I don't think stability is a concern; the sleep is not for asynchronous activity to complete but to allow for the passage of time

Making the test faster definitely a good idea and a mocking functions that the sweeper scheduler functions uses to get time would be the way to go.

nickelization commented on test/riak_kv_sweeper_SUITE.erl:
```diff
@@ -207,16 +309,62 @@ test_scheduler_worker_process_crashed(Config) ->
 
     sweep_all_indices(Indices),
-    timer:sleep(min_scheduler_response_time_msecs()),
+    timer:sleep(timer:seconds(RunIntervalFun() + 1)),
```
Something that just occurred to me: we have a couple tests now that verify that sweeps run after the run interval has expired, but do we have any tests that verify the inverse? We may also want to test that sweeps _don't_ run again until the run interval has elapsed. Perhaps if we could rework the tests to get rid of the sleep, then maybe that's a test we could add at the same time (e.g. if we use something like wait_until, we could track how long it had to wait, and make sure it had to wait at least as long as the run_interval).

raghavkarol replied:
That's a good point, I'll add a test for it.

nickelization commented:
@raghavkarol Added a few review comments, but they're not necessarily super related to this PR. If you want to address them here, great, but if not let me know and I can +1 this so that we can fix the other stuff in separate PRs.

Commit:
Fix failed_sweep function

It is a sweeper callback function that was modified in 781d5b9.

Commit:
Add tests for bucket props

Commit:
Corrected  spelling ObjecSizeBytes -> ObjectSizeBytes

Commit:
Remove `timer:sleep(...)` and `sweep_participant_testing_expire`

Using a mock instead to control advancing of time in from the
testcases. In addition to making the tests faster the code handling of
the `sweep_participant_testing_expire` as `immediate`, only used for
testing is also removed.

Commit:
Add test case for run_interval not expired

Test that sweeps are not re-scheduled before their `run_interval`
time.

raghavkarol commented:
@nickelization - removed the timer sleep from the tests and added a cases for checking that sweeps are not scheduled before their `run_interval`, please review.

Commit:
Fix dialyzer error after changes in 769d063

raghavkarol commented:
@nickelization - removed the timer sleep from the tests and added a cases for checking that sweeps are not scheduled before their run_interval, please review.

nickelization commented:
Awesome, looks great @raghavkarol!

nickelization commented:
+1
